llvm/utils/TableGen/DirectiveEmitter.cpp

```diff
@@ -231,6 +244,8 @@ static void EmitDirectivesDecl(RecordKeeper &Records, raw_ostream &OS) {
   OS << "bool isAllowedClauseForDirective(Directive D, "
      << "Clause C, unsigned Version);\n";
   OS << "\n";
+  OS << "const llvm::SmallVector<Directive> &getLeafConstructs(Directive D);\n";
```

Comment:
```suggestion
  OS << "ArrayRef<Directive> getLeafConstructs(Directive D);\n";
```

Reply:
I don't think this is a good idea.  Returning a reference is pretty much free, Returning an ArrayRef will always create a temporary object.  You can always use ArrayRef in the caller, if that's what's desired there.

Reply:
No, it won't. ArrayRef is not an owning container, so it does not create a copy.

Reply:
The ArrayRef will be the temporary object.

Reply:
No, it is not.

Reply:
I'm expecting you to be more professional in your replies.  I'm simply trying to understand your motivation here.

Reply:
Read the documentation https://llvm.org/docs/ProgrammersManual.html#llvm-adt-arrayref-h

Reply:
That's about input parameters (because then we don't restrict what objects can be passed, i.e. converted to an ArrayRef).  For return values it doesn't matter, at least from the point of view of interface flexibility.  The static constructor issue you pointed out will need this though.